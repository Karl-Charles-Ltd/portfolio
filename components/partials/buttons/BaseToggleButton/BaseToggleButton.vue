<template>
  <div class="v-toggle-btn">
    <label v-if="$slots.default" class="v-toggle-btn__label" :for="id">
      <slot />
    </label>
    <button
      :id="id"
      :aria-checked="on"
      class="v-toggle-btn__pill"
      role="switch"
      v-bind="$attrs"
      @click="toggle"
    >
      <span class="v-toggle-btn__toggle" :class="{ 'v-toggle-btn__toggle--active': on }" />
    </button>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop } from 'nuxt-property-decorator';

@Component({
  inheritAttrs: false,
})
export default class BaseToggleButton extends Vue {
  @Prop({ required: true }) readonly id!: string;

  on: boolean = false;

  toggle() {
    this.on = !this.on;
    this.$emit('toggled', this.on);
  }
}
</script>

<style lang="scss">
.v-toggle-btn {
  &__pill {
    -webkit-appearance: none !important;
    -moz-appearance: none !important;
    appearance: none !important;
    background-color: color('light-grey-one');
    border: none;
    box-sizing: border-box;
    height: $spacing-tiny;
    margin: 0 $spacing-mini;
    outline: none;
    position: relative;
    width: $spacing-small;

    &:before {
      left: -50%;
    }

    &:after {
      right: -50%;
    }
  }

  .v-toggle-btn__pill:before,
  .v-toggle-btn__pill:after {
    background-color: color('light-grey-one');
    content: '';
    height: $spacing-tiny;
    width: $spacing-tiny;
  }

  .v-toggle-btn__pill:before,
  .v-toggle-btn__pill:after,
  .v-toggle-btn__toggle {
    border-radius: 50%;
    position: absolute;
    right: -50%;
    top: 50%;
    transform: translateY(-50%);
  }

  &__toggle {
    background-color: color('white');
    border: solid 1px color('light-grey-one');
    height: $spacing-small;
    left: -#{rem-value(10)};
    transition: all 0.1s ease-in-out;
    width: $spacing-small;
    z-index: z-index('componentsForeground');

    &--active {
      background-color: color('secondary');
      border: solid 1px color('secondary');
      left: rem-value(8);
    }
  }
}
</style>
