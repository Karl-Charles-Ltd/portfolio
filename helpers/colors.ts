/**
 * Get the hex color codes that are used throughout the project.
 * These colors and names should match with
 * assets/scss/settings/_settings.colors.scss.
 *
 * @param {String} colorName - The color name from the theme
 * @returns {String} - The hex code for the matching color
 */
export function colorFromTheme(colorName) {
  const colors = {
    primary: '#6b2f93',
    secondary: '#D0438F',
    tertiary: '#09ECFD',
    quaternary: '#f40ee3',
    quinary: '#370b3c',
    black: '#000000',
    charcoal: '#4D4D4D',
    'grey-one': '#949494',
    'grey-two': '#CDCDCD',
    'light-grey': '#E2E2E2',
    'light-grey-one': '#E2E2E2',
    'light-grey-two': '#F9F9F9',
    white: '#FFFFFF',
  };

  return colors[colorName];
}

export const themeMap = {
  primary: {
    background: 'primary',
    foreground: 'secondary',
  },
  'primary-dark': {
    background: 'primary-dark',
    foreground: 'secondary',
  },
  secondary: {
    background: 'secondary',
    foreground: 'white',
    accent: 'tertiary',
  },
  'secondary-dark': {
    background: 'secondary-dark',
    foreground: 'white',
  },
  'secondary-light': {
    background: 'secondary-light',
    foreground: 'white',
  },
  tertiary: {
    background: 'tertiary',
    foreground: 'secondary',
  },
  quaternary: {
    background: 'quaternary',
    foreground: 'secondary',
  },
  quinary: {
    background: 'quinary',
    foreground: 'white',
  },
  'quinary-darker': {
    background: 'quinary-darker',
    foreground: 'white',
  },
  white: {
    background: 'white',
    foreground: 'secondary',
  },
  charcoal: {
    background: 'charcoal',
    foreground: 'white',
  },
  'light-grey-one': {
    background: 'light-grey-one',
    foreground: 'secondary',
  },
};
