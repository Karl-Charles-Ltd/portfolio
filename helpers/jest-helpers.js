import { shallowMount, createLocalVue } from '@vue/test-utils';
import { createRenderer } from 'vue-server-renderer';
import VueRouter from 'vue-router';

export function setup() {
  const renderer = createRenderer();
  const localVue = createLocalVue();
  const router = new VueRouter();

  // setupPlugins(useVuex, localVue);

  return {
    renderer,
    router,
    localVue,
    shallowMount,
  };
}

export function componentFactory(component, options) {
  return shallowMount(component, options);
}

export function createSnapshot(component, options) {
  const wrapper = componentFactory(component, options);
  expect(wrapper.exists()).toBeTruthy();
  expect(wrapper.html()).toMatchSnapshot();
}
