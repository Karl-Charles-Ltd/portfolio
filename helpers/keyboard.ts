export function addKeyupEventListener(keyCode, callBack) {
  if (!window) return;
  window.addEventListener('keyup', (event) => {
    if (event.keyCode === keyCode) callBack();
  });
}

export function removeKeyupEventListener(keyCode, callBack) {
  if (!window) return;
  window.removeEventListener('keyup', (event) => {
    if (event.keyCode === keyCode) callBack();
  });
}
