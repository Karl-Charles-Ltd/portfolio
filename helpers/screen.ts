import * as screenBreakpoints from '~/constants/screen-breakpoints';

export const breakpoints: string[] = [
  screenBreakpoints.MOBILE,
  screenBreakpoints.TABLET,
  screenBreakpoints.SMALL_DESK,
  screenBreakpoints.LARGE_DESK,
  screenBreakpoints.HUGE_DESK,
];

interface Screen {
  [key: string]: string | null;
}

export default function mq({ from = null, until = null }: Screen, currentScreenSize: string): boolean {
  const currentBreakpointIndex = breakpoints.indexOf(currentScreenSize);

  if (from !== '' && from !== null) {
    const fromBreakpointIndex = breakpoints.indexOf(from || '');
    if (fromBreakpointIndex > currentBreakpointIndex) {
      return false;
    }
  }

  if (until !== '' && until !== null) {
    const untilBreakpointIndex = breakpoints.indexOf(until || '');
    if (untilBreakpointIndex <= currentBreakpointIndex) {
      return false;
    }
  }

  return true;
}
