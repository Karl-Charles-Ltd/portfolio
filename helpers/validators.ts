import { themeMap } from './colors';

export function validateTheme(theme: string, extraColors: string[] = []): boolean {
  const baseThemes: string[] = Object.keys(themeMap);
  const availableThemes: string[] = baseThemes.concat(extraColors);

  return availableThemes.includes(theme);
}
