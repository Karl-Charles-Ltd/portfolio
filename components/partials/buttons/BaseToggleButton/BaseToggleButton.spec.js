import BaseToggleButton from './BaseToggleButton.vue';
import { setup, createSnapshot, componentFactory } from '~/helpers/jest-helpers';

const { localVue } = setup();
const shallowMountComponent = (options) => componentFactory(BaseToggleButton, { localVue, ...options });

describe('BaseToggleButton', () => {
  let propsData;
  let slots;
  beforeEach(() => {
    propsData = {
      id: 'test',
    };
  });
  it('is a Vue instance', () => {
    createSnapshot(BaseToggleButton, { localVue, propsData, slots });
  });
  describe('methods', () => {
    describe('toggle', () => {
      it('sets the "on" boolean true', () => {
        const wrapper = shallowMountComponent({ propsData });
        wrapper.vm.toggle();
        expect(wrapper.vm.on).toBeTruthy();
      });
      it('emits "on" data property', () => {
        const wrapper = shallowMountComponent({ propsData });
        wrapper.vm.toggle();
        expect(wrapper.emitted('toggled')[0][0]).toBeTruthy();
      });
    });
  });
});
