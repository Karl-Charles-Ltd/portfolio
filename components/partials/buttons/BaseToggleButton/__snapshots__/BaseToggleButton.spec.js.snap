// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseToggleButton is a Vue instance 1`] = `
"<div class=\\"v-toggle-btn\\">
  <!----> <button id=\\"test\\" role=\\"switch\\" class=\\"v-toggle-btn__pill\\"><span class=\\"v-toggle-btn__toggle\\"></span></button></div>"
`;
